### PR TITLE
add transform_object_properties in jerryx/arg

### PR DESCRIPTION
**Pull request by jiangzidong**
Add a function in jerryx/arg.
```
jerry_value_t jerryx_arg_transform_object_properties (const jerry_value_t taregt,
                                                      const jerry_char_t **name_p,
                                                      const jerry_length_t name_cnt,
                                                      const jerryx_arg_t *c_arg_p,
                                                      jerry_length_t c_arg_cnt);
```
It will validate the properties of a JS object (by given the prop name list), and convert those properties into native type.

JerryScript-DCO-1.0-Signed-off-by: Zidong Jiang zidong.jiang@intel.com

### CODE REVIEW

File: docs/09.EXT-REFERENCE-ARG.md [hunk 1]
```diff
@@ -206,6 +206,36 @@ jerryx_arg_transform_args (const jerry_value_t *js_arg_p,
...
+
+**Summary**
+
+Validate the properties of a target JS object and assign them to the native arguments.
+
+**Prototype**
+
+```c
+jerry_value_t
+jerryx_arg_transform_object_properties (const jerry_value_t taregt,
```
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: typo: taregt -> target

Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: second thought: is 'target' the best name for this parameter? it is too generic, a bit non-descriptive, e.g., it does not reveal that the function expects an object here.

File: jerry-ext/include/jerryscript-ext/arg.h [hunk 2]
```diff
@@ -64,6 +64,12 @@ jerry_value_t jerryx_arg_transform_args (const jerry_value_t *js_arg_p,
                                          const jerryx_arg_t *c_arg_p,
                                          jerry_length_t c_arg_cnt);
 
+jerry_value_t jerryx_arg_transform_object_properties (const jerry_value_t taregt,
```
Comment (akosthekiss) [inline jerry-ext/include/jerryscript-ext/arg.h]: typo and/or better name here, too

File: tests/unit-ext/test-ext-arg.c [hunk 3]
```diff
@@ -211,6 +215,44 @@ test_validator2_handler (const jerry_value_t func_obj_val __attribute__((unused)
...
+  const char *name_p[] = {"prop1", "prop2", "prop3"};
+
+  jerryx_arg_t mapping[] =
+  {
+    jerryx_arg_boolean (&native1, JERRYX_ARG_COERCE, JERRYX_ARG_REQUIRED),
+    jerryx_arg_number (&native2, JERRYX_ARG_COERCE, JERRYX_ARG_REQUIRED),
+    jerryx_arg_number (&native3, JERRYX_ARG_COERCE, JERRYX_ARG_OPTIONAL)
+  };
+
+  jerry_value_t is_ok = jerryx_arg_transform_object_properties (args_p[0],
```
Comment (akosthekiss) [inline tests/unit-ext/test-ext-arg.c]: how does the newly introduced function (`jerryx_arg_transform_object_properties`) collaborate with the existing solutions (`jerryx_arg_transform_args` and `jerryx_arg_transform_this_and_args`)? if this `test_validator_prop` had more than one arguments - some expected to be objects, some expected to be primitive values - would we have to run one of the `_transform_args` functions first, then "manually" the `_transform_object_properties` for each object argument? is there no way to integrate the two approaches into a single step (from user perspective)? (for now, I'm just asking. I haven't taken a closer look whether at whether it is possible or not.)

File: jerry-ext/include/jerryscript-ext/arg.h [hunk 4]
```diff
@@ -64,6 +64,12 @@ jerry_value_t jerryx_arg_transform_args (const jerry_value_t *js_arg_p,
...
+                                                      const jerry_char_t **name_p,
+                                                      const jerry_length_t name_cnt,
+                                                      const jerryx_arg_t *c_arg_p,
+                                                      jerry_length_t c_arg_cnt);
```
Comment (zherczeg) [inline jerry-ext/include/jerryscript-ext/arg.h]: Do we need two counters? (name_cnt and c_arg_cnt)? Why one of them is const and the other is not? Btw const is not needed for non pointer types.

File: docs/09.EXT-REFERENCE-ARG.md (hunk 1 above)
Comment (jiangzidong) [inline docs/09.EXT-REFERENCE-ARG.md]: what about `obj_val`?

File: jerry-ext/include/jerryscript-ext/arg.h (hunk 4 above)
Comment (jiangzidong) [inline jerry-ext/include/jerryscript-ext/arg.h]: > Do we need two counters?

Yes, just like in `jerryx_arg_transform_args`, we have both js_arg_cnt and c_arg_cnt.
Because name_p and c_arg_p are not 1:1 mappings (depends on the numbers of `jerryx_arg_js_iterator_pop` and `jerryx_arg_js_iterator_peek` used in each validator function).

> Why one of them is const and the other is not?

because `c_arg_cnt`'s value will change in the function call. see line41 in `jerryx_arg_transform_args`: we have `c_arg_cnt--`

> const is not needed for non pointer types.

aligned with other functions' signature in arg.c. And also I think it is a good habit to explicitly declare some vars as const, so that some IDE can help us find the error if we changed its value by mistake, right?

File: tests/unit-ext/test-ext-arg.c (hunk 3 above)
Comment (jiangzidong) [inline tests/unit-ext/test-ext-arg.c]: This is my usage of `jerryx_arg_transform_object_properties` in our internal project:

* create a custom validate function in the top-layer `jerryx_arg_transform_this_and_args` for the js object
* in the custome validator, use `jerryx_arg_transform_object_properties` to check & transform all its properties

File: docs/09.EXT-REFERENCE-ARG.md (hunk 1 above)
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: works with me

File: tests/unit-ext/test-ext-arg.c (hunk 3 above)
Comment (akosthekiss) [inline tests/unit-ext/test-ext-arg.c]: let me try and pick your mind: what else do those custom validator do? do they always just call `jerryx_arg_transform_object_properties`? couldn't we have this function as
```c
jerry_value_t
jerryx_arg_transform_object_properties (jerryx_arg_js_iterator_t *js_arg_iter_p, 
                                        const jerryx_arg_t *c_arg_p) 
```
and have`name_p`, `name_cnt`, `c_arg_p`, `c_arg_cnt` in `c_arg_p->extra_info`?

Comment (jiangzidong) [inline tests/unit-ext/test-ext-arg.c]: The custom validator will prepare the name_p, mappings, and other arguments needed, and then call the `jerryx_arg_transform_object_properties`. 
Because the custom function is prepared for one certain object, for example, `var color = {r:255, g:0, b:0, a:100}`, and this kind of object will used in many binding functions. In my ways, all those binding functions just need to call the custom_function instead of preparing name_p and mapping multiple times in each function bodies.

Comment (akosthekiss) [inline tests/unit-ext/test-ext-arg.c]: But in that case, where are the native variables corresponding to `r`, `g`, `b`, and `a`? Are they global? I cannot see another way of returning their values to the function that calls `jerryx_arg_transform_this_and_args`. If they are global, that will make the transform function useful in serial use cases only.

Comment (jiangzidong) [inline tests/unit-ext/test-ext-arg.c]: My example:
in binding.c, the function handler which use color
```c
typedef struct
{
uint8 red;
uint8 green,
uint8 blue;
uint8 alpha
} native_color_t;

jerry_value_t
set_color_handler(
	const jerry_value_t	func_obj_val,
	const jerry_value_t	this_val,
	const jerry_value_t	args[],
	const jerry_length_t	 args_cnt)
{
	native_view_t *view_p;
	native_color_t color;

	jerryx_arg_t mapping[] =
	{
		jerryx_arg_native_pointer(
			(void **)&view_p,
			&JERRY_NATIVE_HANDLE_INFO_FOR_CTYPE(view),
			JERRYX_ARG_REQUIRED
			),
		jerryx_arg_custom(&color, 0, js_arg_check_is_color)
	};

	jerry_value_t is_ok = jerryx_arg_transform_args(args, args_cnt, mapping,
							2);
      ....
```

and the js_arg_check_is_color is my custom function
```c
jerry_value_t
js_arg_check_is_color(jerryx_arg_js_iterator_t *js_arg_iter_p,
		const jerryx_arg_t *	c_arg_p)
{
	jerry_value_t js_arg = jerryx_arg_js_iterator_pop (js_arg_iter_p);
	const char *keys[] = {"r", "g", "b", "a"};
	native_color_t *color_p = c_arg_p->dest;

	const jerryx_arg_t mapping[] = {
	jerryx_arg_uint8 (&color_p->red, JERRYX_ARG_COERCE, JERRYX_ARG_REQUIRED),
	jerryx_arg_uint8 (&color_p->green, JERRYX_ARG_COERCE, JERRYX_ARG_REQUIRED),
	jerryx_arg_uint8 (&color_p->blue, JERRYX_ARG_COERCE, JERRYX_ARG_REQUIRED),
	jerryx_arg_uint8 (&color_p->alpha, JERRYX_ARG_COERCE, JERRYX_ARG_REQUIRED),
	};
	
	return jerryx_arg_transform_object_properties(js_arg, keys, 4, mapping, 4);
}
```

Comment (akosthekiss) [inline tests/unit-ext/test-ext-arg.c]: I see, it works well if you have to extract properties from the same type of objects over and over again. It might be a burden though, to define a custom transform function, if your object argument is used in a single handler only. I wonder if it might be useful to have something like this available to the user:

```c
typedef struct
{
  const jerry_char_t **name_p,   /**< property name list of the JS object */
  const jerry_length_t name_cnt, /**< count of the name list */
  const jerryx_arg_t *c_arg_p,   /**< points to the array of transformation steps */
  const jerry_length_t c_arg_cnt /**< the count of the `c_arg_p` array */
} jerryx_arg_object_props_t;

jerry_value_t
jerryx_arg_transform_object_props (jerryx_arg_js_iterator_t *js_arg_iter_p,
                                   const jerryx_arg_t *c_arg_p)
{
  jerry_value_t js_arg = jerryx_arg_js_iterator_pop (js_arg_iter_p);
  
  const jerryx_arg_object_props_t *object_props = (const jerryx_arg_object_props_t *) c_arg_p->extra_info;

  return jerryx_arg_transform_object_properties (js_arg, object_props->name_p, object_props->name_cnt, object_props->c_arg_p, object_props->c_arg_cnt);
} /* jerryx_arg_transform_object_props */

static inline jerryx_arg_t
jerryx_arg_object_properties (const jerryx_arg_object_props_t *object_props) /**< pointer to object property mapping */
{
  return (jerryx_arg_t)
  {
    .func = jerryx_arg_transform_object_props,
    .dest = NULL,
    .extra_info = object_props
  };
} /* jerryx_arg_object_properties */
```

It may tempt users to use this function instead of writing specialised custom transform functions, of course.

File: jerry-ext/arg/arg.c [hunk 5]
```diff
@@ -81,8 +81,45 @@ jerryx_arg_transform_this_and_args (const jerry_value_t this_val, /**< the this_
...
+  {
+    return jerry_create_error (JERRY_ERROR_TYPE, (jerry_char_t *) "Not an object.");
+  }
+
+  jerry_value_t prop[name_cnt];
+
+  for (jerry_length_t i = 0; i < name_cnt; i++, name_p++)
+  {
+    const jerry_value_t name_str = jerry_create_string (*name_p);
+    prop[i] = jerry_get_property (target, name_str);
```
Comment (martijnthe) [inline jerry-ext/arg/arg.c]: Need to check for error here.
What if the property does not exist?

Comment (jiangzidong) [inline jerry-ext/arg/arg.c]: Ok, because the jerry_get_property may call the getter function, and the getter may throw error. Otherwise, it cannot be error.

> What if the property does not exist?

Then prop[i] is undefined, it will pass the the validator if its corresponding native arg is set as "optional".

File: docs/09.EXT-REFERENCE-ARG.md [hunk 6]
```diff
@@ -31,6 +31,31 @@ typedef struct
...
 - [jerryx_arg_native_pointer](#jerryx_arg_native_pointer)
 - [jerryx_arg_ignore](#jerryx_arg_ignore)
+- [jerryx_arg_object_properties](#jerryx_arg_object_properties)
+
+## jerryx_arg_object_props_t
+
+**Summary**
+
+The structure is used in `jerryx_arg_object_properties`. It provides the properties' names,
+its corresponding JS-to-C mapping and other related infomation.
```
Comment (LaszloLango) [inline docs/09.EXT-REFERENCE-ARG.md]: typo: `information`

File: docs/09.EXT-REFERENCE-ARG.md [hunk 7]
```diff
@@ -31,6 +31,31 @@ typedef struct
...
+
+```c
+typedef struct
+{
+  const jerry_char_t **name_p; /**< property name list of the JS object */
+  jerry_length_t name_cnt; /**< count of the name list */
+  const jerryx_arg_t *c_arg_p; /**< points to the array of transformation steps */
+  jerry_length_t c_arg_cnt; /**< the count of the `c_arg_p` array */
+} jerryx_arg_object_props_t;
+
```
Comment (LaszloLango) [inline docs/09.EXT-REFERENCE-ARG.md]: remove this empty line

File: docs/09.EXT-REFERENCE-ARG.md [hunk 8]
```diff
@@ -335,6 +394,91 @@ jerryx_arg_native_pointer (void **dest,
...
+  double required_num;
+  double optional_num = 1234.567;  // default value
+
+  // "prop_name_p" defines the name list of the expected properties' names
+  const char *prop_name_p[] = {"enable", "data", "extra_data"}
+  }
+
+  // "prop_mapping" defines the steps to transform properties to C variables:
+  const jerryx_arg_t prop_mapping[] =
+    jerryx_arg_boolean (&required_bool, JERRYX_ARG_COERCE, JERRYX_ARG_REQUIRED),
```
Comment (LaszloLango) [inline docs/09.EXT-REFERENCE-ARG.md]: missing open brace

File: docs/09.EXT-REFERENCE-ARG.md [hunk 9]
```diff
@@ -335,6 +394,91 @@ jerryx_arg_native_pointer (void **dest,
...
+                                          const jerry_value_t this_val,
+                                          const jerry_value_t args_p[],
+                                          const jerry_length_t args_count)
+{
+  bool required_bool;
+  double required_num;
+  double optional_num = 1234.567;  // default value
+
+  // "prop_name_p" defines the name list of the expected properties' names
+  const char *prop_name_p[] = {"enable", "data", "extra_data"}
```
Comment (LaszloLango) [inline docs/09.EXT-REFERENCE-ARG.md]: wrong indentation and missing space after `{` and before `}`

File: docs/09.EXT-REFERENCE-ARG.md [hunk 10]
```diff
@@ -335,6 +394,91 @@ jerryx_arg_native_pointer (void **dest,
...
+    .name_p = (const jerry_char_t **) prop_name_p,
+    .name_cnt = ARRAY_LENGTH (prop_name_p),
+    .c_arg_p = prop_mapping,
+    .c_arg_cnt = ARRAY_LENGTH (prop_mapping)
+  };
+
+  // It is the mapping used in the jerryx_arg_transform_this_and_args
+  const jerryx_arg_t mapping[] =
+  {
+    jerryx_arg_object_properties (&prop_info),
```
Comment (LaszloLango) [inline docs/09.EXT-REFERENCE-ARG.md]: unnecessary `,`

File: jerry-ext/arg/arg.c [hunk 11]
```diff
@@ -81,8 +81,50 @@ jerryx_arg_transform_this_and_args (const jerry_value_t this_val, /**< the this_
...
+    return jerry_create_error (JERRY_ERROR_TYPE, (jerry_char_t *) "'this' validation failed.");
   }
 
   return jerryx_arg_transform_args (js_arg_p, js_arg_cnt, c_arg_p + 1, c_arg_cnt - 1);
 } /* jerryx_arg_transform_this_and_args */
+
+/**
+ * Validate the `obj_val`'s properties,
+ * and assign them to the native arguments.
+ * @return jerry undefined: all validators passed,
```
Comment (LaszloLango) [inline jerry-ext/arg/arg.c]: missing newline before `@return`

File: jerry-ext/include/jerryscript-ext/arg.impl.h [hunk 12]
```diff
@@ -275,4 +277,20 @@ jerryx_arg_custom (void *dest, /**< pointer to the native argument where the res
...
+ * Create a jerryx_arg_t instance for object properties.
+ *
+ * @return a jerryx_arg_t instance.
+ */
+static inline jerryx_arg_t
+jerryx_arg_object_properties (const jerryx_arg_object_props_t *object_props) /**< pointer to object property mapping */
+{
+  return (jerryx_arg_t)
+  {
+  .func = jerryx_arg_transform_object_props,
```
Comment (LaszloLango) [inline jerry-ext/include/jerryscript-ext/arg.impl.h]: wrong indentation

**Comment by jiangzidong** [comment]
@LaszloLango updated

File: tests/unit-ext/test-ext-arg.c (hunk 3 above)
Comment (martijnthe) [inline tests/unit-ext/test-ext-arg.c]: @akosthekiss I like your suggestion / addition.

This is useful when implementing a native handler that takes arguments plus an additional `options` object alongside the other arguments. Example:
https://developer.mozilla.org/en-US/docs/Web/API/Geolocation/watchPosition

File: docs/09.EXT-REFERENCE-ARG.md [hunk 13]
```diff
@@ -206,6 +231,39 @@ jerryx_arg_transform_args (const jerry_value_t *js_arg_p,
 - [jerryx_arg_transform_this_and_args](#jerryx_arg_transform_this_and_args)
 
 
+## jerryx_arg_transform_object_properties
+
+**Summary**
+
+Validate the properties of a target JS object and assign them to the native arguments.
```
Comment (martijnthe) [inline docs/09.EXT-REFERENCE-ARG.md]: I'd remove the word "target", I think.
When I read this, it makes me think we're setting something on the "target" / we're "targeting" the JS object.

File: jerry-ext/arg/arg.c [hunk 14]
```diff
@@ -81,8 +81,51 @@ jerryx_arg_transform_this_and_args (const jerry_value_t this_val, /**< the this_
...
+  jerry_value_t prop[name_cnt];
+
+  for (jerry_length_t i = 0; i < name_cnt; i++, name_p++)
+  {
+    const jerry_value_t name_str = jerry_create_string (*name_p);
+    prop[i] = jerry_get_property (obj_val, name_str);
+    jerry_release_value (name_str);
+
+    if (jerry_value_has_error_flag (prop[i]))
+    {
```
Comment (martijnthe) [inline jerry-ext/arg/arg.c]: If an error happened at `i == N`, you need to release `prop[0..N-1]` before returning.

Comment (martijnthe) [inline jerry-ext/arg/arg.c]: Add a unit test case that covers this path?

File: tests/unit-ext/test-ext-arg.c (hunk 3 above)
Comment (jiangzidong) [inline tests/unit-ext/test-ext-arg.c]: Yes, I added the `jerryx_arg_object_properties ` and also kept the `jerryx_arg_transform_object_properties` @akosthekiss @martijnthe

File: jerry-ext/arg/arg.c (hunk 14 above)
Comment (jiangzidong) [inline jerry-ext/arg/arg.c]: Yes! Forgot the other props...

Comment (jiangzidong) [inline jerry-ext/arg/arg.c]: done

**Comment by jiangzidong** [comment]
Further comments? @LaszloLango @akosthekiss @zherczeg @martijnthe

**Comment by zherczeg** [comment]
I am not deeply involved in argument processing so I cannot say much.

File: docs/09.EXT-REFERENCE-ARG.md [hunk 15]
```diff
@@ -206,6 +231,39 @@ jerryx_arg_transform_args (const jerry_value_t *js_arg_p,
...
 
 
+## jerryx_arg_transform_object_properties
+
+**Summary**
+
+Validate the properties of a JS object and assign them to the native arguments.
+
+*Note*: User can call it directly to transform the object properties, or they can use
+`jerryx_arg_object_properties` in `jerryx_arg_transform_this_and_args` or `jerryx_arg_transform_args`.
```
Comment (martijnthe) [inline docs/09.EXT-REFERENCE-ARG.md]: Perhaps a bit more clear:

> *Note*: This function transforms properties of a *single* JS object into native C values. To transform *multiple* objects in one pass – for example when converting multiple arguments to an external handler – use `jerryx_arg_object_properties` together with `jerryx_arg_transform_this_and_args` or `jerryx_arg_transform_args`

File: docs/09.EXT-REFERENCE-ARG.md [hunk 16]
```diff
@@ -335,6 +393,89 @@ jerryx_arg_native_pointer (void **dest,
...
+                                                      mapping,
+                                                      ARRAY_LENGTH (mapping));
+
+  if (jerry_value_has_error_flag (rv))
+  {
+    // Handle error
+    return rv;
+  }
+
+  // Validated and tranformed successfully!
```
Comment (martijnthe) [inline docs/09.EXT-REFERENCE-ARG.md]: Typo: transformed (missing s)

File: jerry-ext/arg/arg.c [hunk 17]
```diff
@@ -81,8 +81,56 @@ jerryx_arg_transform_this_and_args (const jerry_value_t this_val, /**< the this_
...
+      for (jerry_length_t j = 0; j < i; j++)
+      {
+        jerry_release_value (prop[j]);
+      }
+
+      return prop[i];
+    }
+  }
+
+  jerry_value_t ret = jerryx_arg_transform_args (prop, name_cnt, c_arg_p, c_arg_cnt);
```
Comment (martijnthe) [inline jerry-ext/arg/arg.c]: `const jerry_value_t ret` ?

File: jerry-ext/arg/arg-transform-functions.c [hunk 18]
```diff
@@ -270,6 +270,27 @@ jerryx_arg_transform_native_pointer (jerryx_arg_js_iterator_t *js_arg_iter_p, /*
...
 
 /**
+ * Check whether the JS object's properties have expected types, and transform them into native args.
+ *
+ * @return jerry undefined: the transformer passes,
+ *         jerry error: the transformer fails.
+ */
+jerry_value_t
+jerryx_arg_transform_object_props (jerryx_arg_js_iterator_t *js_arg_iter_p,
+                                   const jerryx_arg_t *c_arg_p)
```
Comment (LaszloLango) [inline jerry-ext/arg/arg-transform-functions.c]: Missing comments for the arguments

File: docs/09.EXT-REFERENCE-ARG.md [hunk 19]
```diff
@@ -335,6 +393,89 @@ jerryx_arg_native_pointer (void **dest,
...
+static jerry_value_t my_external_handler (const jerry_value_t function_obj,
+                                          const jerry_value_t this_val,
+                                          const jerry_value_t args_p[],
+                                          const jerry_length_t args_count)
+{
+  bool required_bool;
+  double required_num;
+  double optional_num = 1234.567;  // default value
+
+  // "prop_name_p" defines the name list of the expected properties' names
```
Comment (LaszloLango) [inline docs/09.EXT-REFERENCE-ARG.md]: Please use multi-line comments `/* */`

**Comment by jiangzidong** [comment]
Fixed @martijnthe @LaszloLango

**Comment by jiangzidong** [comment]
@akosthekiss Any comments?

File: docs/09.EXT-REFERENCE-ARG.md [hunk 20]
```diff
@@ -206,6 +231,41 @@ jerryx_arg_transform_args (const jerry_value_t *js_arg_p,
...
 
 
+## jerryx_arg_transform_object_properties
+
+**Summary**
+
+Validate the properties of a JS object and assign them to the native arguments.
+
+*Note*: This function transforms properties of a single JS object into native C values. 
+To transform multiple objects in one pass(for example when converting multiple arguments
```
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: missing space at "pass(for"

File: docs/09.EXT-REFERENCE-ARG.md [hunk 21]
```diff
@@ -335,6 +395,90 @@ jerryx_arg_native_pointer (void **dest,
 
 - [jerryx_arg_transform_this_and_args](#jerryx_arg_transform_this_and_args)
 
+## jerryx_arg_object_properties
+
+**Summary**
+
+Create a validation/transformation step (`jerryx_arg_t`) that expects to
+consume one `object` JS argument and call `jerryx_arg_transform_object_properties` inside
```
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: `object` or `Object`? (`jerryx_arg_native_pointer` uses the capitalized form. I have no preference but for uniformity)

File: docs/09.EXT-REFERENCE-ARG.md [hunk 22]
```diff
@@ -335,6 +395,90 @@ jerryx_arg_native_pointer (void **dest,
...
+    jerryx_arg_boolean (&required_bool, JERRYX_ARG_COERCE, JERRYX_ARG_REQUIRED),
+    jerryx_arg_number (&required_num, JERRYX_ARG_COERCE, JERRYX_ARG_REQUIRED),
+    jerryx_arg_number (&optional_num, JERRYX_ARG_COERCE, JERRYX_ARG_OPTIONAL)
+  };
+
+  /* Prepare the jerryx_arg_object_props_t instance. */
+  const jerryx_arg_object_props_t prop_info =
+  {
+    .name_p = (const jerry_char_t **) prop_name_p,
+    .name_cnt = ARRAY_LENGTH (prop_name_p),
```
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: Is this a standard macro somewhere? I acknowledge that it's intuitive, but I don't know where to look for its definition

(Just realizing that it already appears in this .md file in the example of `jerryx_arg_transform_this_and_args`. I'm not sure that it should have appeared there either.)

File: docs/09.EXT-REFERENCE-ARG.md (hunk 21 above)
Comment (jiangzidong) [inline docs/09.EXT-REFERENCE-ARG.md]: I will change to `object`

File: docs/09.EXT-REFERENCE-ARG.md (hunk 22 above)
Comment (jiangzidong) [inline docs/09.EXT-REFERENCE-ARG.md]: I will remove the ARRAY_LENGTH and use the literal number directly.

The example here is different from that in jerryx_arg_transform_this_and_args. This example is showing how to check object's properties using `jerryx_arg_transform_args`

File: docs/09.EXT-REFERENCE-ARG.md [hunk 23]
```diff
@@ -138,33 +162,33 @@ static jerry_value_t my_external_handler (const jerry_value_t function_obj,
   char required_str[16];
   double optional_num = 1234.567;  // default value
 
-  // "mapping" defines the steps to transform input arguments to C variables:
+  /* "mapping" defines the steps to transform input arguments to C variables. */
   const jerryx_arg_t mapping[] =
   {
-    // `this` is the first value. No checking needed on `this` for this function.
+    /*`this` is the first value. No checking needed on `this` for this function. */
```
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: a space got lost after `/*`

File: docs/09.EXT-REFERENCE-ARG.md [hunk 24]
```diff
@@ -138,33 +162,33 @@ static jerry_value_t my_external_handler (const jerry_value_t function_obj,
...
-    // Handle error
+    /* Handle error. */
     return rv;
   }
 
-  // Validated and tranformed successfully!
-  // required_bool, required_str and optional_num can now be used.
-  // ...
+  /* Validated and transformed successfully!
+   * required_bool, required_num and optional_num can now be used. */
```
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: why was this changed? the variable name is still `required_str`

File: docs/09.EXT-REFERENCE-ARG.md [hunk 25]
```diff
@@ -314,7 +374,7 @@ jerryx_arg_function (jerry_value_t *dest,
 **Summary**
 
 Create a validation/transformation step (`jerryx_arg_t`) that expects to
-consume one `Object` JS argument that is 'backed' with a native pointer with
+consume one `object` JS argument that is 'backed' with a native pointer with
```
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: the doc should be updated in the source as well

File: jerry-ext/include/jerryscript-ext/arg.h [hunk 26]
```diff
@@ -105,6 +121,8 @@ static inline jerryx_arg_t
 jerryx_arg_ignore (void);
 static inline jerryx_arg_t
 jerryx_arg_custom (void *dest, uintptr_t extra_info, jerryx_arg_transform_func_t func);
+static inline jerryx_arg_t
+jerryx_arg_object_properties (const jerryx_arg_object_props_t *object_props_p);
```
Comment (akosthekiss) [inline jerry-ext/include/jerryscript-ext/arg.h]: could object args not be optional like most of the args, i.e., what about adding `jerryx_arg_optional_t opt_flag` to the arg list of this function?

File: docs/09.EXT-REFERENCE-ARG.md (hunk 24 above)
Comment (jiangzidong) [inline docs/09.EXT-REFERENCE-ARG.md]: According to this comments https://github.com/jerryscript-project/jerryscript/pull/1879#discussion_r121090770
We should use `/* */` instead of `//`

File: docs/09.EXT-REFERENCE-ARG.md [hunk 27]
```diff
@@ -138,33 +162,35 @@ static jerry_value_t my_external_handler (const jerry_value_t function_obj,
...
+    /* Handle error. */
     return rv;
   }
 
-  // Validated and tranformed successfully!
-  // required_bool, required_str and optional_num can now be used.
-  // ...
+  /*
+   * Validated and transformed successfully!
+   * required_bool, required_num and optional_num can now be used.
```
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: There is a misunderstanding. The way the comment was formatted (with `/* ... */`) was OK with me. The problem is that the comment text itself has changed. Somehow required_str became required_num.

File: docs/09.EXT-REFERENCE-ARG.md [hunk 28]
```diff
@@ -335,6 +397,93 @@ jerryx_arg_native_pointer (void **dest,
...
+Create a validation/transformation step (`jerryx_arg_t`) that expects to
+consume one `object` JS argument and call `jerryx_arg_transform_object_properties` inside
+to transform its properties to native arguments.
+User should prepare the `jerryx_arg_object_props_t` instance, and pass it to this function.
+
+**Prototype**
+
+```c
+static inline jerryx_arg_t
+jerryx_arg_object_properties (const jerryx_arg_object_props_t *object_props_p);
```
Comment (akosthekiss) [inline docs/09.EXT-REFERENCE-ARG.md]: opt_flag is documented below but does not appear in the signature

**Comment by jiangzidong** [comment]
updated @akosthekiss
